Finish halfbrained's awesome work

Akuli opened the pull request:
- [x] Figure out when hover responses can be null, add test: created #45

Akuli commented on sansio_lsp_client/structs.py:
```diff
@@ -51,11 +54,15 @@ class VersionedTextDocumentIdentifier(TextDocumentIdentifier):
     version: t.Optional[int]
 
 
+# Sorting tip:  sorted(positions, key=(lambda p: (p.line, p.character)))
```
```suggestion
# Sorting tip:  sorted(positions, key=(lambda p: p.as_tuple()))
```

Akuli commented on sansio_lsp_client/structs.py:
```diff
@@ -203,19 +260,189 @@ class DiagnosticSeverity(enum.IntEnum):
     INFORMATION = 3
     HINT = 4
 
+    # TODO: delete this? it's easy to implement in your editor if you really need it
+    def short_name(self) -> str:
+        return {
+            self.ERROR: "Err",
+            self.WARNING: "Wrn",
+            self.INFORMATION: "Inf",
+            self.HINT: "Hint",
+        }[self]
```
@halfbrained What do you think about this?

halfbrained replied:
No objections

Akuli commented on sansio_lsp_client/structs.py:
```diff
@@ -51,11 +54,15 @@ class VersionedTextDocumentIdentifier(TextDocumentIdentifier):
     version: t.Optional[int]
 
 
+# Sorting tip:  sorted(positions, key=(lambda p: (p.line, p.character)))
 class Position(BaseModel):
     # NB: These are both zero-based.
     line: int
     character: int
 
+    def as_tuple(self) -> t.Tuple[int, int]:
+        return (self.line, self.character)
```
@halfbrained My `as_tuple()` method replaces `__lt__` in this class and `pos()` in `class DocumentSymbol`. What do you think?

Akuli replied:
After making this review comment, I changed the lambda to `lambda p: p.as_tuple()`, which is even simpler.

halfbrained replied:
Is there a benefit to this change? :)

Akuli replied:
I think `Position.__lt__()` and `DocumentSymbol.pos()` were both weird:
- If there is `__lt__`, why are `__gt__`, `__le__` and `__ge__` missing? It's because `__lt__` is what `sort()` and `sorted()` use, but it's not obvious when looking at the code. We could add a comment about it, but `key=(lambda p: p.as_tuple())` is very obvious as is.
- If you see `something.pos()` in the middle of longer code, you wouldn't expect it to take just the **start** of a range and ignore the end, which is what that method did. On the other hand, if you see `something.selectionRange.start.as_tuple()`, you immediately know it uses just the start position and the end is being ignored.

Akuli commented on sansio_lsp_client/structs.py:
```diff
@@ -192,6 +242,13 @@ class Location(BaseModel):
 
+class LocationLink(BaseModel):
+    originSelectionRange: t.Optional[Range]
+    targetUri: str  # DocumentUri...
```
@halfbrained What does this comment mean?

halfbrained replied:
Just a note that in the specification this field is of type `DocumentUri`

Akuli replied:
```suggestion
    targetUri: str  # in the spec the type is DocumentUri
```

Akuli commented on sansio_lsp_client/structs.py:
```diff
@@ -203,19 +260,189 @@ class DiagnosticSeverity(enum.IntEnum):
+    INCREMENTAL = 2
+
+
+class SymbolInformation(BaseModel):  # symbols: flat list
+    name: str
+    kind: SymbolKind
+    tags: t.Optional[t.List[SymbolTag]]
+    deprecated: t.Optional[bool]
+    location: Location
+    containerName: t.Optional[str]
+
+
+class DocumentSymbol(BaseModel):  # symbols: hierarchy
```
@halfbrained What do the `# symbols: ...` comments mean?

halfbrained replied:
Meaning is: `SymbolInformation` is just a flat list of symbols, `DocumentSymbol` - a tree - `METHOD` items are children of `CLASS` items, etc. 

It's not clear from just the type names.

Akuli commented on tests/test_actual_langservers.py:
```diff
@@ -0,0 +1,551 @@
+    # threaded
+    def _read_loop(self):
+        try:
+            while self._pout:
```
```suggestion
            while True:
```

Akuli commented on tests/test_actual_langservers.py:
```diff
@@ -0,0 +1,551 @@
+    # threaded
+    def _send_loop(self):
+        try:
+            while self._pin:
```
```suggestion
            while True:
```

Akuli commented on tests/test_actual_langservers.py:
```diff
@@ -0,0 +1,551 @@
+        self._read_data_received()
+
+    def do_method(self, text, file_uri, method, response_type=None):
+        def doc_pos():  # SKIP
```
@halfbrained What doees this comment mean?

halfbrained replied:
It's for my editor :) no other purpose

Akuli commented:
Marking as draft because I split some of this to separate pull requests that should be merged first.

PurpleMyst commented:
Hey @Akuli, I saw you've requested my review on a couple PRs. I've been *very* busy with an upcoming exam. I'll be free in the incoming week, so I'll take a look at them then. Sorry for the delay.

Akuli commented:
Awesome! Let's review the other PRs next week.

Akuli commented on tests/test_actual_langservers.py:
```diff
@@ -106,12 +132,26 @@ def _read_data_received(self):
     def wait_for_message_of_type(self, type_, timeout=5):
-        end_time = time.monotonic() + timeout
-        while time.monotonic() < end_time:
+        end_time = time.time() + timeout
+        while time.time() < end_time:
```
Merge problem

```suggestion
        end_time = time.monotonic() + timeout
        while time.monotonic() < end_time:
```

Akuli commented on tests/test_actual_langservers.py:
```diff
@@ -322,8 +379,7 @@ def do_bar():
                 "Use of undeclared identifier 'do_'",
-                "Expected '}'\n\nfoo.c:9:16: note: to match this '{'",
-                "To match this '{'\n\nfoo.c:9:37: error: expected '}'",
+                "Expected '}'",
```
No idea why this changes

halfbrained replied:
If I recall correctly - that's a result of sending client's capabilities during the server initialization

Akuli commented on tests/test_actual_langservers.py:
```diff
@@ -349,6 +406,101 @@ def do_bar():
+            assert typedef.result[0].uri == (project_root / filename).as_uri()
+
+
+def _needs_clangd(version):
+    return pytest.mark.skipif(
+        not list(langserver_dir.glob(f"clangd_{version}.*")),
+        reason=f"clangd {version} not found",
+    )
+
 
```
Not used

```suggestion
```

Akuli commented:
@PurpleMyst Ready for review

PurpleMyst commented on sansio_lsp_client/client.py:
```diff
@@ -171,14 +329,49 @@ def parse_request(event_cls: t.Type[Event]) -> Event:
+        elif request.method == "window/workDoneProgress/create":
+            assert request.params is not None
+            event = parse_request(WorkDoneProgressCreate)
+            self._progress_tokens_map[request.params["token"]] = WorkDoneProgress
```
Why is this a `dict`? It seems to only have `WorkDoneProgress` as a value

Akuli replied:
I cleaned it up. I think halfbrained's original idea was to implement more stuff later, but we can easily add this back if it's needed.

PurpleMyst commented on sansio_lsp_client/client.py:
```diff
@@ -211,7 +410,7 @@ def shutdown(self) -> None:
     def exit(self) -> None:
         assert self._state == ClientState.SHUTDOWN
-        self._send_notification(method="exit")
+        self._send_notification(method="exit", params={})
```
I'd like it if we could put a comment here explaining why we can't just not specify `params`

Akuli replied:
I thought this was another gopls thing, but changing it to `params=None` doesn't break gopls tests.

Akuli replied:
I added a TODO comment about this

halfbrained replied:
It was indeed added in gopls fixes, but I don't remember the events leading up to it, and I already deleted all the go stuff, sorry.

PurpleMyst commented on sansio_lsp_client/client.py:
```diff
@@ -285,3 +495,111 @@ def completions(
+    def definition(
+        self,
+        text_document_position: TextDocumentPosition,
+        # TODO PartialResultParams
```
What does this comment mean?

Akuli replied:
I don't know, but if it's something important, we'll discover it later. I deleted the comments for now. They really aren't of any value when nobody understands them.

PurpleMyst commented on sansio_lsp_client/events.py:
```diff
@@ -78,3 +130,98 @@ class WillSaveWaitUntilEdits(Event):
     diagnostics: t.List[Diagnostic]
+
+
+class Hover(Event):
```
What do the comments in this class body mean?

Akuli replied:
They listed attributes that the members had, which is just duplicate information, because you can go to where the member type is defined and look. I deleted them.

PurpleMyst commented on sansio_lsp_client/structs.py:
```diff
@@ -51,11 +54,15 @@ class VersionedTextDocumentIdentifier(TextDocumentIdentifier):
     version: t.Optional[int]
 
 
+# Sorting tip:  sorted(positions, key=(lambda p: p.as_tuple()))
```
Should we add some comparison operators here so that this is redundant?

Akuli replied:
There used to be comparison operators (actually just `__lt__`), but I switched to this because it seemed a lot simpler. Imagine seeing this in the middle of longer code:
- `sorted(positions)` is like "Can you really do this? Does it work? Is this a typo?"
-  `sorted(positions, key=...)` is like "We compare them as tuples, so it works"

Before `.as_tuple()`, there was also a convenience method that returned `foo.range.start.as_tuple()` (I think), and that was replaced by `.as_tuple()` too. Overall `.as_tuple()` feels like a nice and minimal approach to me.

PurpleMyst replied:
Understandable.

PurpleMyst reviewed (changes requested):
LGTM but I'm still iffy about the whole logging business. Perhaps we could do it a different way? Don't feel great about having a hidden switch just for testing.

Akuli commented:
I deleted the logging stuff. Its purpose was to allow receiving events even if receiving one of them errors, so that you still get the other events. I changed `recv()` to return a generator that yields correctly parsed messages and then fails, but doesn't leave the failed message to `_recv_buf` so that it will succeed next time. I feel like it's almost too elegant; the code doesn't look tricky, even though I have specifically written it so that it handles many different cases.

Akuli commented:
@halfbrained, can you explain your comments to us for one last time? :)

Akuli commented:
@PurpleMyst Ready for review again

PurpleMyst commented:
LGTM!

halfbrained commented:
> @halfbrained, can you explain your comments to us for one last time? :)

My apologies for missing this.

(Not sure how that happened, but I'm almost certain that's github having a laugh at my expense... :) )
